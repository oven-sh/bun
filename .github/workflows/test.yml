name: Test

concurrency:
  group: ${{ github.workflow }}-${{ inputs.run-id || github.run_id }}
  cancel-in-progress: true

on:
  push:
  workflow_call:
    inputs:
      run-id:
        type: string
        required: true

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          run-id: ${{ inputs.run-id || '8623134412' }}
          pattern: bun-*
      - run: |
          ls -la
