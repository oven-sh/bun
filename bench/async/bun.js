import { bench, run } from "mitata";

bench("sync", () => {});
bench("async", async () => {});
bench("await 1", async () => await 1);

await run();
