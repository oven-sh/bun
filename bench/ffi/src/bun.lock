{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@napi-rs/cli": "^2.10.1",
        "@node-rs/helper": "^1.3.3",
      },
    },
  },
  "packages": {
    "@napi-rs/cli": ["@napi-rs/cli@2.14.8", "", { "bin": { "napi": "scripts/index.js" } }, "sha512-IvA3s8BqohMdUbOkFn7+23u1dhIJZCkA8Xps7DD4SLdCMbcbUF6MUuKiqxuqmVHBFTaxU25sU56WdX3efqGgPw=="],

    "@napi-rs/triples": ["@napi-rs/triples@1.1.0", "", {}, "sha512-XQr74QaLeMiqhStEhLn1im9EOMnkypp7MZOwQhGzqp2Weu5eQJbpPxWxixxlYRKWPOmJjsk6qYfYH9kq43yc2w=="],

    "@node-rs/helper": ["@node-rs/helper@1.3.3", "", { "dependencies": { "@napi-rs/triples": "^1.1.0" } }, "sha512-p4OdfQObGN9YFy5WZaGwlPYICQSe7xZYyXB0sxREmvj1HzGKp5bPg2PlfgfMZEfnjIA882B9ZrnagYzZihIwjA=="],
  }
}
