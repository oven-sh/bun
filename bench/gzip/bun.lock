{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@babel/standalone": "7.24.10",
      },
    },
  },
  "packages": {
    "@babel/standalone": ["@babel/standalone@7.24.10", "", {}, "sha512-nGC37EKfmelpyCXto1pw6SBkD5ZQRdMbL6WISi28xWit9dtiy9dChU1WgEfzturUTxrmOGkMDRrCydFMA7uOaQ=="],
  }
}
