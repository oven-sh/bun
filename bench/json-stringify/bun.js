import { bench, run } from "mitata";

bench("JSON.stringify({hello: 'world'})", () =>
  JSON.stringify({ hello: "world" }),
);

const otherUint8Array = new Uint8Array(1024);
bench("Uint8Array.from(otherUint8Array)", () =>
  Uint8Array.from(otherUint8Array),
);

run();
