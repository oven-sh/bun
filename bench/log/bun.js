import { bench, run } from "mitata";

bench("console.log('hello')", () => console.log("hello"));
bench("console.log({ hello: 'object' })", () => console.log({ hello: "object" }));
await run();
