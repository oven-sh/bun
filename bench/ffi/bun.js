import { ptr, dlopen, CString, toBuffer } from "bun:ffi";
import { run, bench, group } from "mitata";

const { napiNoop, napiHash, napiString } = require(import.meta.dir + "/src/ffi_napi_bench.node");

const {
  symbols: {
    ffi_noop: { native: ffi_noop },
    ffi_hash: { native: ffi_hash },
    ffi_string: { native: ffi_string },
  },
} = dlopen(import.meta.dir + "/src/ffi_napi_bench.node", {
  ffi_noop: { args: [], returns: "void" },
  ffi_string: { args: [], returns: "ptr" },
  ffi_hash: { args: ["ptr", "u32"], returns: "u32" },
});

const bytes = new Uint8Array(64);

group("bun:ffi", () => {
  bench("noop", () => ffi_noop());
  bench("hash", () => ffi_hash(ptr(bytes), bytes.byteLength));

  bench("c string", () => new CString(ffi_string()));
});

if (process.env.SHOW_NAPI)
  group("bun:napi", () => {
    bench("noop", () => napiNoop());
    bench("hash", () => napiHash(bytes));

    bench("string", () => napiString());
  });

await run();
