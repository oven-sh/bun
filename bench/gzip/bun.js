import { run, bench, group } from "mitata";
import { gzipSync, gunzipSync } from "bun";

const data = new TextEncoder().encode("Hello World!".repeat(9999));

const compressed = gzipSync(data);

const libraries = ["zlib"];
if (Bun.semver.satisfies(Bun.version, ">=1.1.21")) {
  libraries.push("libdeflate");
}

const options = { library: undefined };
const benchFn = (name, fn) => {
  if (libraries.length > 1) {
    group(name, () => {
      for (const library of libraries) {
        bench(library, () => {
          options.library = library;
          fn();
        });
      }
    });
  } else {
    bench(name, () => {
      fn();
    });
  }
};

benchFn(`roundtrip - "Hello World!".repeat(9999))`, () => {
  gunzipSync(gzipSync(data, options), options);
});

benchFn(`gzipSync("Hello World!".repeat(9999)))`, () => {
  gzipSync(data, options);
});

benchFn(`gunzipSync("Hello World!".repeat(9999)))`, () => {
  gunzipSync(compressed, options);
});

await run();
