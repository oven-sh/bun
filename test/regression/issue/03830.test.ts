// test/regression/issue/03830.test.ts

import { it, expect } from "bun:test";
import { bunEnv, bunExe } from "harness";
import { mkdirSync, rmSync, writeFileSync, readFileSync, mkdtempSync } from "fs";
import { tmpdir } from "os";
import { join } from "path";

it("macros should not lead to seg faults under any given input", async () => {
  // this test code follows the same structure as and
  // is based on the code for testing issue 4893

  const testDir = mkdtempSync(join(tmpdir(), "issue3830-"));

  // Clean up from prior runs if necessary
  rmSync(testDir, { recursive: true, force: true });

  // Create a directory with our test file
  mkdirSync(testDir, { recursive: true });
  writeFileSync(join(testDir, "macro.ts"), "export function fn(str) { return str; }");
  writeFileSync(join(testDir, "index.ts"), "import { fn } from './macro' assert { type: 'macro' };\nfn(`Â©${''}`);");

  const { stdout, exitCode } = Bun.spawnSync({
    cmd: [bunExe(), "build", join(testDir, "index.ts")],
    env: bunEnv,
    stderr: "inherit",
  });

  expect(exitCode).toBe(0);
});
