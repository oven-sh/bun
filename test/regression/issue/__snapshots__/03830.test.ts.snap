// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`macros should not lead to seg faults under any given input 1`] = `
"2 | fn(\`©${''}\`);
    ^
error: "Cannot convert argument type to JS" error in macro
    at [dir]/index.ts:2:1"
`;

exports[`macros should not lead to seg faults under any given input 1`] = `
"2 | fn(\`©${''}\`);
    ^
error: "Cannot convert argument type to JS" error in macro
    at [dir]/index.ts:2:1"
`;

exports[`macros should not lead to seg faults under any given input 1`] = `
"2 | fn(\`©${''}\`);
    ^
error: "Cannot convert argument type to JS" error in macro
    at [dir]/index.ts:2:1"
`;

exports[`macros should not lead to seg faults under any given input 1`] = `
"2 | fn(\`©${''}\`);
    ^
error: "Cannot convert argument type to JS" error in macro
    at [dir]/index.ts:2:1"
`;
